packages/react/src/use-action-form/utils.ts

```diff
@@ -460,6 +460,23 @@ export function applyDataMask(opts: {
 
   const readOnlyPaths = getReadOnlyPaths(select, modelApiIdentifier);
 
+  // Check for relationship fields that are entirely read-only, including nested relationships
+  const checkRelationshipFields = (obj: any, prefix = "") => {
+    const relationshipFields = Object.keys(obj || {}).filter((key) => typeof obj[key] === "object" && obj[key] !== null);
+    for (const field of relationshipFields) {
+      const currentPrefix = prefix ? `${prefix}.${field}` : field;
+      const fieldPaths = readOnlyPaths.filter((path) => path.startsWith(`${currentPrefix}`));
+      if (fieldPaths.length > 0 && fieldPaths.length === Object.keys(obj[field]).length) {
+        readOnlyPaths.push(currentPrefix);
+      } else {
+        // Recursively check nested relationships
+        checkRelationshipFields(obj[field], currentPrefix);
+      }
+    }
+  };
+
+  checkRelationshipFields(select, modelApiIdentifier);
+
```

Comment:
`readOnlyPaths` ends up as a list of paths like `answer,question.text`. which would be the key of a flattened object. Perhaps we could get around the thing I observed in the tests with a flattened object approach